Data/ByteString/Lazy.hs

```diff
@@ -608,11 +623,47 @@ scanl
     -- ^ input of length n
     -> ByteString
     -- ^ output of length n+1
-scanl f z = snd . foldl k (z,singleton z)
- where
-    k (c,acc) a = let n = f c a in (n, acc `snoc` n)
+scanl f = flip (foldr go singleton)
+  where
+    go value continuation accumulator =
+      let next = f accumulator value
+      in accumulator `cons` continuation next
 {-# INLINE scanl #-}
 
+-- | 'scanl1' is a variant of 'scanl' that has no starting value argument.
+-- This function will fuse.
+--
+-- > scanl1 f [x1, x2, ...] == [x1, x1 `f` x2, ...]
+scanl1 :: (Word8 -> Word8 -> Word8) -> ByteString -> ByteString
+scanl1 _ Empty = Empty
+scanl1 f (Chunk c cs) = scanl f (S.unsafeHead c) (chunk (S.unsafeTail c) cs)
+
+-- | 'scanr' is similar to 'foldr', but returns a list of successive
+-- reduced values from the right.
+--
+-- > scanr f z [..., x{n-1}, xn] == [..., x{n-1} `f` (xn `f` z), xn `f` z, z]
+--
+-- Note that
+--
+-- > head (scanr f z xs) == foldr f z xs
+-- > last (scanr f z xs) == z
+--
+scanr
+    :: (Word8 -> Word8 -> Word8)
+    -- ^ element -> accumulator -> new accumulator
+    -> Word8
+    -- ^ starting value of accumulator
+    -> ByteString
+    -- ^ input of length n
+    -> ByteString
+    -- ^ output of length n+1
+scanr f z = pack . fmap (foldr f z) . tails
```

Comment:
This is a bit unfortunate, because there is no sharing between `tails`, so you end up folding each tail with `f` independently, O(n^2) operations in total.

Reply:
Yes, I would like to do something about it. I need to meditate on this. Can I maybe reverse the thing and then fold it from the start? It is forced all along anyway.

I am not sure how to think about this sort of things.

Reply:
I have not gave it much thought, but maybe we can start from a high-level picture? Let's write lazy 
```haskell
mapAccumLChunks :: (acc -> Strict.ByteString -> (acc, Strict.ByteString)) -> acc -> Lazy.ByteString -> (acc, ByteString)

mapAccumRChunks :: (acc -> Strict.ByteString -> (acc, Strict.ByteString)) -> acc -> Lazy.ByteString -> (acc, ByteString)
```
Then reuse them in definitions of `mapAccum{L,R}`, and `scan{l,r}` as well (`scans` are just a special case of `mapAccum`).

Reply:
My understanding is that, since we output a byte string about as long as the input, time and space complexity is bounded below by _O(n)_. This means to my mind that reversing and then using `scanl` is as efficient as it gets. Is there anything I get wrong?

Reply:
I think we can do a bit better than reversing the entire input. I think we can do one pass to reverse the order of the _chunks_ and determine the total length.  Then do a second pass over the reversed sequence of chunks and write the output.

Benchmarks would be useful to check that we're roughly in the right ballpark.

Reply:
Alright, I get it like this:

* We agree that linear time and space are lower bounds.
* We still want to avoid some expensive operations, like reversing all chunks.

Sounds about right?

Reply:
Yep, sound right to me! :)